cram/cram_io.c

```diff
@@ -1901,326 +1903,109 @@ int cram_set_header(cram_fd *fd, SAM_hdr *hdr) {
 
-	if (-1 == refs2id(fd->refs, fd->header))
-	    return -1;
-	if (!fd->refs->ref_id || !fd->refs->ref_id[id])
-	    return -1;
+        if (-1 == refs2id(fd->refs, fd->header))
+            return -1;
+        if (!fd->refs->ref_id || !fd->refs->ref_id[id])
+            return -1;
 
-	// Local copy already, so fall back to cram_get_ref().
-	return 0;
+        // Local copy already, so fall back to cram_get_ref().
+        return 0;
     }
-
-    /* Populate the local disk cache if required */
-    if (local_cache && *local_cache) {
-        int pid = (int) getpid();
-        unsigned thrid = get_int_threadid();
-        hFILE *fp;
-
-        if (*cache_root && !is_directory(cache_root)) {
-            hts_log_warning("Creating reference cache directory %s\n"
-                "This may become large; see the samtools(1) manual page REF_CACHE discussion",
-                cache_root);
+    else{
+        // Got the ref from m5_to_ref, set the relevant properties for the caller
+        hFILE_ref *hfref = (hFILE_ref*)ref;
+
+        r->length = hfref->length;
+        r->is_md5 = 1;
+
+        if(!hfref->is_mem_hfile){
+            r->offset = r->line_length = r->bases_per_line = 0;
+            // set the fp and fn to the value returned by m5_to_ref
+            r->fn = string_dup(fd->refs->pool, hfref->file_name);
+            if (fd->refs->fp)
+                if (bgzf_close(fd->refs->fp) != 0)
+                    return -1;
+            fd->refs->fp = bgzf_hopen(ref, "r");
+            fd->refs->fn = r->fn;
         }
+        else{
+            char* seq;
+            size_t seq_sz;
+            if((seq = hfile_mem_get_buffer(hfref->base.innerhf, &seq_sz)) == NULL)
+                return -1;
+            free(ref); // Can't hclose as this would free the buffer
 
-	expand_cache_path(path, local_cache, tag->str+3);
-	hts_log_info("Writing cache file '%s'", path);
-	mkdir_prefix(path, 01777);
-
-	do {
-	    // Attempt to further uniquify the temporary filename
-	    unsigned t = ((unsigned) time(NULL)) ^ ((unsigned) clock());
-	    thrid++; // Ensure filename changes even if time/clock haven't
-
-	    sprintf(path_tmp, "%s.tmp_%d_%u_%u", path, pid, thrid, t);
-	    fp = hopen(path_tmp, "wx");
-	} while (fp == NULL && errno == EEXIST);
-	if (!fp) {
-	    perror(path_tmp);
-
-	    // Not fatal - we have the data already so keep going.
-	    return 0;
-	}
-
-	// Check md5sum
-	hts_md5_context *md5;
-	char unsigned md5_buf1[16];
-	char md5_buf2[33];
-
-	if (!(md5 = hts_md5_init())) {
-	    hclose_abruptly(fp);
-	    unlink(path_tmp);
-	    return -1;
-	}
-	hts_md5_update(md5, r->seq, r->length);
-	hts_md5_final(md5_buf1, md5);
-	hts_md5_destroy(md5);
-	hts_md5_hex(md5_buf2, md5_buf1);
-
-	if (strncmp(tag->str+3, md5_buf2, 32) != 0) {
-	    hts_log_error("Mismatching md5sum for downloaded reference");
-	    hclose_abruptly(fp);
-	    unlink(path_tmp);
-	    return -1;
-	}
+            r->seq = seq;
+            r->mf = hfref->mf;
```

Comment:
As we have `hfref=ref` above on line 1981 and `free(ref)` at line 2001, this is accessing freed memory.  The free statement needs moving down a few lines.

Reply:
Furthermore, the `free(ref)` isn't sufficient.  If ref is an hFILE_wrapper then we also need to free the innerhf too. Eg:

```
            if (ref->is_hFILE_wrapper)
                free(((hFILE_wrapper *)ref)->innerhf); // hack!                                                  
            free(ref); // Can't hclose as this would free the buffer                                             
```

This is all feeling like the wrong solution and it's going to fail if we ever have other forms of wrappers, but without the above we leak memory.

I'm trying to recall the reasoning for the wrapper and why it can't just be an `hFILE_mem` hfile class without any wrapping up.  I think the wrapper is there because we have another type (`hFILE_ref`), but perhaps we just want hFILE_mem expanded to be able to do the same as the ref one.

The ref version adds mf and file_name fields.  mf is only needed due to mmap and the requirement to unmap at the end.  File_name I can't see a use for in this code.   If we have requirements for these then we may be better off adding them both to an extended hFILE_mem structure instead of introducing an entirely new wrapper layer.

Reply:
@jkbonfield The use of hfile_wrappers comes from the need to use the existing code in cram_populate_ref and functions that call it, without much modification - I didn't want to rewrite the logic in `cram_populate_ref` and `cram_get_ref` (and maybe other places). This required a structure that could hold both in-memory structures to be contained as well holding arbitrary hFiles generated from https://github.com/wtsi-hgi/htslib/blob/1a6a7e242d75c9b746653ee548d2bce430cb905b/ref.c#L304. The file name, length and mFile (if used) also needed to be stored, where the file name and length are used here: https://github.com/wtsi-hgi/htslib/blob/1a6a7e242d75c9b746653ee548d2bce430cb905b/cram/cram_io.c#L1989 and here: https://github.com/wtsi-hgi/htslib/blob/1a6a7e242d75c9b746653ee548d2bce430cb905b/cram/cram_io.c#L1983. 

I could rewrite relevant code such that `m5_to_ref` would use an extended in-memory hFile pointing either memory or mmapped data, where this extended structure would contain a flag indicating whether this is an mmapped structure and it's length. However, it may be better to work to merge this PR as it is with the inelegant implementation, then refractor the code at a later time.